### PR TITLE
Midi 50/data preprocessing pipeline

### PR DESCRIPTION
**Pull request by WojciechMat**
[MIDI translation transformer](https://sacrebleu.atlassian.net/jira/software/projects/MIDI/boards/7?selectedIssue=MIDI-50)

I have used @JasiekKaczmarczyk 's code for quantization to start working on data pipeline for transformer which will generate durations and velocities, given a sequence of pitches and low-resolution dstart.

It does not do what it is supposed to do yet, work in progress.

### CODE REVIEW

**Comment by WojciechMat** [comment]
## TokenizedMidiDataset
I have implemented a dataset, which tokenizes midi files from maestro dataset. You can see an example by running:
```
python dataset.py
```
 There is no need for padding right now, as the sequences are always the same lenght. 
Is the way I do this okay?
Should I add something more?

File: data/tokenizer.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+        for idx in range(record[self.keys[0]].size):
+            sample = str()
+
+            for key in self.keys:
+                sample += f"{record[key][idx]}-"
+
+            samples.append(sample[:-1])
+
+        return samples
```
Comment (roszcz) [inline data/tokenizer.py]:
```suggestion
    def __call__(self, df: pd.DataFrame) -> list[str]:
        samples = []
        
        for it, row in df.iterrows():
            sample = "-".join([row[key] for key in self.keys])
            samples.append(samples)
        return samples
```

It's good to think about midi structures as data frames. If you don't want to use pandas here for speed reasons, you should still try to simplify this method and try to make it look/read more like this pandas version :)

File: data/dataset.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+            eos_id = torch.tensor([1], device=self.device)  # </s> token id
+            src = torch.cat(
+                [bs_id, torch.tensor(src_processed, dtype=torch.int64, device=self.device), eos_id],
+                0,
+            )
+            tgt = torch.cat(
+                [bs_id, torch.tensor(tgt_processed, dtype=torch.int64, device=self.device), eos_id],
+                0,
+            )
+
```
Comment (roszcz) [inline data/dataset.py]: I think those cats are not necessary - if you want to add the bos/eos tokens,  maybe a better place to do this would be in the tokenizer class, so later you can just do `tgt = torch.tensor(tht_processed)`.

P.S. One could argue, that for a problem with constant sequence length bos/eos tokens are not needed, but I guess we can keep them - ingesting custom tokens is a good exercise :)

File: data/dataset.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        self.src_vocab, self.tgt_vocab = self._build_vocab()
+
+        self.processed_records = []
+        self.samples = []
+
+        self._build()
+
+    def _build(self):
+        self.processed_records = self.load_dataset()
+
```
Comment (roszcz) [inline data/dataset.py]:
```suggestion
        self.processed_records = self.load_dataset()
        self.samples = self._build()

    def _build(self) -> list[Sample]:
        ...
```

I think it's better to be explicit about property creation in the `__init__`, than to hide it in methods

**Review by roszcz — COMMENTED** [review]
Thanks, looks great!

One project level thing to fix is the path management - looks like you're running the code from the `dataset` folder, but I think we should do it from the project root in this case, so instead of `python dataset.py` -> `python -m data.dataset` or something like that

File: data/dataset.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        self.device = device
+
+        self.quantizer = MidiQuantizer(
+            n_dstart_bins=n_dstart_bins,
+            n_duration_bins=n_duration_bins,
+            n_velocity_bins=n_velocity_bins,
+        )
+
+        self.tokenizer_src = Tokenizer(keys=["pitch", "dstart_bin"])
+        self.tokenizer_tgt = Tokenizer(keys=["duration_bin", "velocity_bin"])
```
Comment (roszcz) [inline data/dataset.py]: My initial idea was about predicting a high resolution velocity information, from low resolution quantization:

```py
# 3 * 3 * 3 * 88 = 2376 tokens
self.quantizer = MidiQuantizer(
    n_dstart_bins=3,
    n_duration_bins=3,
    n_velocity_bins=3,
)

# Source tokenizer should use all (low resolution) information
self.tokenizer_src = Tokenizer(keys=["pitch", "dstart_bin", "duration_bin", "velocity_bin"])

# This can have up to 128 tokens, which would be one token per every possible velocity value
self.tokenizer_tgt = VelocityTokenizer()
```

So I think it might be better to have dedicated tokenizers for ins and outs of the problem 🤔

Comment (WojciechMat) [reply data/dataset.py]: Do I understand it correctly - we want to have all the quantized information on input and predict only high resolution velocity?
My initial thought was to enable the model to predict higher resolution velocity and duration based on low resolution dstart and pitch. I think it would be interestiong to see waht predictions would the transformer come up with. But we can do velocities first, it will probably work a lot better.

Comment (WojciechMat) [reply data/dataset.py]: + I can still use the same class, I just need to set keys as ['velocity'] and pass original notes df

Comment (roszcz) [reply data/dataset.py]: > Do I understand it correctly - we want to have all the quantized information on input and predict only high resolution velocity?

That's right

> velocities first, it will probably work a lot better

Yes, the idea is to start with an "easy" problem and if that works look for a way to get it to work on more complex problems :)

For the tokenizers, I think having separate classes will be more explicit about what's going on - looks like we'll have to test multiple different approaches to tokenization, and I don't want to build a single class doing everything for this 🤔

Comment (WojciechMat) [reply data/dataset.py]: Okay, I’ll work on it

**Comment by WojciechMat** [comment]
I have implemented your suggestions, I still do not use pandas in tokenization, as it increases computation time like three times.

**Comment by WojciechMat** [comment]
Do you think I should add tokenized dataset saving after tokenization? It can take up to 10 minutes to create one with higher number of bins.

**Comment by roszcz** [comment]
Yes, dataset caching is probably a good idea

**Comment by WojciechMat** [comment]
I have added untokenization method to tokenizer class, dataset caching, \<s> and  \</s> tokens are now added in Tokenizer, and the dataset is oriented towards learning from low resolution, quantized data to predict high resolution velocity.

File: data/dataset.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+        self.dataset = load_dataset(path="roszcz/maestro-v1", split=split)
+
+        self.src_vocab, self.tgt_vocab = self._build_vocab()
+
+        self.processed_records, self.unprocessed_records = self.load_dataset()
+        self.samples = self.load_samples()
+
+    def load_samples(self) -> list[tuple[list[int], list[int]]]:
+        path = (
+            f"data/datasets/samples-"
```
Comment (roszcz) [inline data/dataset.py]:
```suggestion
            f"tmp/datasets/samples-"
```

There's code in the `datasets` folder, and we shouldn't mix storage space with code space :)

**Review by roszcz — APPROVED** [review]
Looks good, ready to 🚂 🚋 🚋
